internal/controller/apiextensions/claim/syncer_ssa.go

```diff
@@ -74,32 +72,76 @@ func NewPatchingManagedFieldsUpgrader(w client.Writer) *PatchingManagedFieldsUpg
 
 // Upgrade the supplied object's field managers from client-side to server-side
 // apply.
-func (u *PatchingManagedFieldsUpgrader) Upgrade(ctx context.Context, obj client.Object, ssaManager string, csaManagers ...string) error {
-	// UpgradeManagedFieldsPatch removes or replaces the specified CSA managers.
-	// Unfortunately most Crossplane controllers use CSA manager "crossplane".
-	// So we could for example fight with the XR controller:
-	//
-	// 1. We remove CSA manager "crossplane", triggering XR controller watch
-	// 2. XR controller uses CSA manager "crossplane", triggering our watch
-	// 3. Back to step 1 :)
-	//
-	// In practice we only need to upgrade once, to ensure we don't share fields
-	// that only this controller has ever applied with "crossplane". We assume
-	// that if our SSA manager already exists, we've done the upgrade.
-	for _, e := range obj.GetManagedFields() {
+//
+// This is a multi-step process.
+//
+// Step 1: All fields are owned by either manager 'crossplane', operation
+// 'Update' or manager 'apiextensions.crossplane.io/composite', operation
+// 'Apply'. This represents all fields set by the claim or XR controller up to
+// this point.
+//
+// Step 2: Upgrade is called for the first time. We delete all field managers.
+//
+// Step 3: The claim controller server-side applies its fully specified intent
+// as field manager 'apiextensions.crossplane.io/claim'. This becomes the
+// manager of all the fields that are part of the claim controller's fully
+// specified intent. All existing fields the claim controller didn't specify
+// become owned by a special manager - 'before-first-apply', operation 'Update'.
+//
+// Step 4: Upgrade is called for the second time. It deletes the
+// 'before-first-apply' field manager entry. Only the claim field manager
+// remains.
+//
+// Step 5: Eventually the XR reconciler updates a field (e.g. spec.resourceRefs)
+// and becomes owner of that field.
+func (u *PatchingManagedFieldsUpgrader) Upgrade(ctx context.Context, obj client.Object, ssaManager string, _ ...string) error {
```

Comment:
I think it's useful to have two implementations (Nop and Patching) because it allows us to use the Nop implementation by default, but use the Patching implementation when the claim SSA feature flag is enabled.

Reply:
Nono, sure, I was talking about the csamanagers argument to the function.

Reply:
I meant this https://github.com/crossplane/crossplane/pull/5641